src/main/java/com/google/cloud/teleport/splunk/SplunkEventWriter.java

```diff
@@ -266,29 +258,32 @@ private void logWriteFailures(@StateId(COUNT_STATE_NAME) ValueState<Long> countS
   }
 
   /**
-   * A helper method to check the {@link HttpResponse} object and wrap the response into an {@link
-   * Optional} {@link SplunkWriteError} in case of an unsuccessful response.
+   * Utility method to un-batch and flush failed write events.
    *
-   * @param response {@link HttpResponse} object
-   * @return {@link Optional} {@link SplunkWriteError}
+   * @param events List of {@link SplunkEvent}s to un-batch
+   * @param statusMessage Status message to be added to {@link SplunkWriteError}
+   * @param statusCode Status code to be added to {@link SplunkWriteError}
+   * @param receiver Receiver to write {@link SplunkWriteError}s to
    */
-  private Optional<SplunkWriteError> checkResponse(HttpResponse response)
-      throws IOException {
-    if (!response.isSuccessStatusCode()) {
-      SplunkWriteError.Builder builder = SplunkWriteError.newBuilder();
-      builder.withStatusCode(response.getStatusCode());
-
-      if (response.getStatusMessage() != null) {
-        builder.withStatusMessage(response.getStatusMessage());
-      }
+  private void flushWriteFailures(
+      List<SplunkEvent> events,
+      String statusMessage,
+      Integer statusCode,
+      OutputReceiver<SplunkWriteError> receiver) {
 
-      if (response.parseAsString() != null) {
-        builder.withPayload(response.parseAsString());
-      }
-      SplunkWriteError error = builder.build();
-      return Optional.of(error);
+    SplunkWriteError.Builder builder = SplunkWriteError.newBuilder();
+    builder.withStatusMessage(statusMessage);
+
+    if (statusCode != null) {
+      builder.withStatusCode(statusCode);
+    }
+
+    for (SplunkEvent event : events) {
```

Comment:
You are assuming `events` would never be null :) If so, better add a checkNotNull() at the beginning of this method as a guardrail?

Reply:
Good catch! I'll add a checkNotNull() at the beginning of this method.